File: 013队列Queue C语言/queue.c
//
// Created by zhubin on 2019/1/27.
//
#include "queue.h"
#include <stdio.h>
#include <string.h>
void initList(Queue* queue){
    queue->head = (ListElmt*)malloc(sizeof(ListElmt));
    queue->tail = NULL;
    queue->size = 0;
}
void queue_destroy(Queue* queue){
    while (queue->size != 0) {
        pop(queue);
    }
    memset(queue,0, sizeof(queue));
}
bool push(Queue* queue,void* data){
    if (queue->size == 0) {
        queue->head->data = data;
        queue->tail = queue->head;
        queue->tail->next = NULL;
    }else{
        ListElmt *new_element = (ListElmt *) malloc(sizeof(ListElmt));
        if (new_element == NULL) {
            puts("内存分配失败");
            return false;
        }
        new_element->data = data;
        queue->tail->next = new_element;
        queue->tail = new_element;
        queue->tail->next = NULL;
    }
    queue->size++;
    return true;
}
bool pop(Queue* queue){
    if (queue->size == 0) {
        puts("栈空!");
        return false;
    }
    if (queue->size == 1) {
        queue->tail = NULL;
    }
    ListElmt *old_element = queue->head;
    queue->head = queue->head->next;
    queue->size--;
    free(old_element);
    return true;

}
void* peek(Queue* queue){
    return  queue->head ==NULL?NULL:queue->head->data;
}

File: 015串 子串匹配算法 Java语言/src/git/src/queue/Queue.java
package git.src.queue;
/*
 * 循环队列
 * 
 * */
public class Queue<T> {
	Object[] array = new Object[3];
	int head = 0;
	int tail = 0;
	int size = 0;
	public Queue() {
	}
	public Queue(int maxSize) {
		array = new Object[maxSize];
	}
	public void push(T data) {
		if (tail == array.length) {
			tail = 0;
		}
		array[tail++] = data;
		size++;
	}
	public Object pop() {
		if (size == 0) {
			try {
				throw new Exception();
			} catch (Exception e) {
				System.err.println("队列已空");
			}
		}
		if(head == array.length) {
			head = 0;
		}
		size--;
		return array[head++];
	}
	public Object peek() {
		return array[head];
	}
	public boolean isEmpty() {
		return size == 0;
	}
	public boolean isFull() {
		return size == array.length;
	}
}
